### PR TITLE
[Geo] Added DeckGL Arc Layer and Refactor on BaseDeckGL class

### LINKED ISSUE #4087: Add support for deck.gl "arc" layer

### PR DESCRIPTION
**Pull request by hughhhh**
- Added `DeckGL.ArcLayer()` layer to be used in Superset
- Refactor on `BaseDeckGL`
   - Created interface model for layers to have more control in there get_data methods
   - Using the control key as reference in the dataframe instead of `lonlat`

screen shot 2017-12-28 at 1 05 25 pm

screen shot 2017-12-28 at 1 04 37 pm

Closes #4087 

@mistercrunch

### CODE REVIEW

File: superset/viz.py [hunk 1]
```diff
@@ -1824,15 +1824,16 @@ def query_obj(self):
...
-        spatial = fd.get('spatial')
-        if spatial:
-            if spatial.get('type') == 'latlong':
-                gb += [spatial.get('lonCol')]
-                gb += [spatial.get('latCol')]
-            elif spatial.get('type') == 'delimited':
-                gb += [spatial.get('lonlatCol')]
-            elif spatial.get('type') == 'geohash':
-                gb += [spatial.get('geohashCol')]
+        for spatial_key in ['spatial', 'start_spatial', 'end_spatial']:
```
Comment (mistercrunch) [inline superset/viz.py]: [optional] could define a `process_spatial_control` method in the base deckgl class, and use it in the subclasses that need it.

**Review by mistercrunch — CHANGES_REQUESTED** [review]
Looks good overall but pointed out a few improvements

**Comment by hughhhh** [comment]
@mistercrunch I added the new `process_spatial` methods, I think its better to do the full refactor in another PR. I used the new methods child Arc class. Let me know what you think!

**Review by mistercrunch — CHANGES_REQUESTED** [review]
Would be great if we got rid of the previous `spatial` control logic as you refactor this.

One way to model it would be to have the derivatives of BaseDeckGL class specify something like `spatial_control_keys = ['spatial']` or  `spatial_control_keys = ['start_spatial', 'end_spatial']` as a static class attribute. Then the `BaseDeckGL.query_obj` method can prepare the query object (prepare the `groupby` key in the dict) and `BaseDeckGL.get_data` can prepare the dataframe (unroll Series in the dataframe). The derived class would just receive the objects already prepared for used.

File: superset/viz.py [hunk 2]
```diff
@@ -1818,6 +1818,58 @@ def get_position(self, d):
...
+            df[cols] = (
+                df[spatial.get('lonlatCol')]
+                .str
+                .split(spatial.get('delimiter'), expand=True)
+                .astype(np.float64)
+            )
+            del df[spatial.get('lonlatCol')]
+        elif spatial.get('type') == 'geohash':
+            latlong = df[spatial.get('geohashCol')].map(geohash.decode)
+            df['lat'] = latlong.apply(lambda x: x[0])
```
Comment (mistercrunch) [inline superset/viz.py]: Idea: it may be less complex overall to simply store the lat/long as a tuple in a single column

File: superset/viz.py [hunk 3]
```diff
@@ -1818,21 +1819,54 @@ def get_position(self, d):
...
+            group_by += [spatial.get('geohashCol')]
+
+    def process_spatial_data_obj(self, key, df):
+        spatial = self.form_data.get(key)
+        if spatial is None:
+            raise ValueError(_('Bad spatial key'))
+
+        if spatial.get('type') == 'latlong':
+            df = df.rename(columns={
+                spatial.get('lonCol'): 'lon',
```
Comment (mistercrunch) [inline superset/viz.py]: Aren't names conflicting here when there are multiple spatial controls? For that reason I think it'd be ideal to use the key as the column name in the dataframe, and store a tuple with lat & long as the value.

**Review by mistercrunch — APPROVED** [review]

**Pull request merged**